compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp

```diff
@@ -0,0 +1,95 @@
+    ArrayRef<OpFoldResult> sizes, ArrayRef<OpFoldResult> strides,
+    ValueRange dynamicDims, SmallVector<OpFoldResult> &clonedOffsets,
+    SmallVector<OpFoldResult> &clonedSizes,
+    SmallVector<OpFoldResult> &clonedStrides,
+    SmallVector<Value> &clonedDynamicDims) {
+  SetVector<Operation *> slice;
+  TransitiveFilter filter = [&nonIndexComputationOperands,
+                             &baseOp](Operation *op) {
+    for (auto val : nonIndexComputationOperands) {
+      if (op == val.getDefiningOp()) return false;
+    }
+    if (op->isProperAncestor(baseOp)) return false;
+    return !isa<IREE::HAL::InterfaceConstantLoadOp>(op);
+  };
+  getBackwardSlice(baseOp, &slice, filter);
+  BlockAndValueMapping bvm;
+  for (auto origOp : slice) {
+    builder.clone(*origOp, bvm);
+  }
+  auto remapOpFoldResult = [&bvm](ArrayRef<OpFoldResult> ofrs) {
+    SmallVector<OpFoldResult> clonedOfrs;
+    clonedOfrs.reserve(ofrs.size());
+    for (auto ofr : ofrs) {
+      if (ofr.is<Attribute>()) {
+        clonedOfrs.push_back(ofr);
+      } else {
+        clonedOfrs.push_back(bvm.lookupOrDefault(ofr.get<Value>()));
+      }
+    }
+    return clonedOfrs;
+  };
+  auto remapValues = [&bvm](ValueRange vals) {
+    SmallVector<Value> clonedVals;
+    clonedVals.reserve(vals.size());
+    for (auto val : vals) {
+      clonedVals.push_back(bvm.lookupOrDefault(val));
+    }
+    return clonedVals;
+  };
+
+  clonedOffsets = remapOpFoldResult(offsets);
+  clonedSizes = remapOpFoldResult(sizes);
+  clonedStrides = remapOpFoldResult(strides);
+  clonedDynamicDims = remapValues(dynamicDims);
+}
+
+void cloneOffsetsSizesAndStrides(OpBuilder &builder,
+                                 IREE::Flow::DispatchTensorStoreOp storeOp,
+                                 SmallVector<OpFoldResult> &offsets,
+                                 SmallVector<OpFoldResult> &sizes,
+                                 SmallVector<OpFoldResult> &strides,
+                                 SmallVector<Value> &dynamicDims) {
+  return cloneOffsetsSizesAndStridesImpl(
+      builder, storeOp, ValueRange{storeOp.getValue(), storeOp.getTarget()},
+      storeOp.getMixedOffsets(), storeOp.getMixedSizes(),
+      storeOp.getMixedStrides(), storeOp.getTargetDims(), offsets, sizes,
+      strides, dynamicDims);
+}
+
+void cloneOffsetsSizesAndStrides(OpBuilder &builder,
+                                 IREE::Flow::DispatchTensorLoadOp loadOp,
+                                 SmallVector<OpFoldResult> &offsets,
+                                 SmallVector<OpFoldResult> &sizes,
+                                 SmallVector<OpFoldResult> &strides,
+                                 SmallVector<Value> &dynamicDims) {
+  return cloneOffsetsSizesAndStridesImpl(
+      builder, loadOp, ValueRange{loadOp.getSource()}, loadOp.getMixedOffsets(),
+      loadOp.getMixedSizes(), loadOp.getMixedStrides(), loadOp.getSourceDims(),
+      offsets, sizes, strides, dynamicDims);
+}
```

Comment:
These function takes tooo many arguments. I think we can use upstream struct `Range` in `mlir/Dialect/Utils/StaticValueUtils.h`.

```cpp
/// Represents a range (offset, size, and stride) where each element of the
/// triple may be dynamic or static.
struct Range {
  OpFoldResult offset;
  OpFoldResult size;
  OpFoldResult stride;
};
```

and we rewrite it to return a `SmallVector<Range>`. I think it is a trivial case for [RVO](https://en.wikipedia.org/wiki/Copy_elision#RVO), and it's cleaner.

```
SmallVector<Range> cloneOffsetsSizesAndStrides(OpBuilder &builder,
  IREE::Flow::DispatchTensorLoadOp loadOp, SmallVector<Value> &dynamicDims).
```

We can have some helper function that construct `SmallVector<Range>` given vector of `sizes`, vector of `offsets` and vector of `strides`.

If we have concerns that `dynamicDims` is passing by reference, we can at least do

```
void cloneOffsetsSizesAndStrides(OpBuilder &builder,
  IREE::Flow::DispatchTensorLoadOp loadOp,
  SmallVector<Range> &ranges, SmallVector<Value> &dynamicDims).
```

Reply:
If you pass `Range`, then you have to first pack everything into a `Range` only to unpack it to get the `SmallVector<OpFoldResult>`  A function with 6 args is not unreasonable. We have `build` methods with 11 arguments, etc.

Reply:
> If you pass Range, then you have to first pack everything into a Range only to unpack it to get the SmallVector<OpFoldResult>

yes, and that could be done by creating a helper function.

> A function with 6 args is not unreasonable. We have build methods with 11 arguments, etc.

This is one of parts that I don't like in MLIR -- a builder methods with many arguments, and the order matters. I'm trying to reduce too many arguments functions when possible. This is what I learnt from https://abseil.io/tips/173

> For functions which take multiple arguments which may be confused by the caller or where you want to specify default arguments without having to worry about the order, strongly consider using option structs to increase both convenience and code clarity.

I'm open to both options, just wanna raise a signal about it and think if we've considered it.